src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_x509.c

```diff
@@ -77,298 +77,119 @@ int32_t AndroidCryptoNative_EncodeX509(jobject /*X509Certificate*/ cert, uint8_t
-    ON_EXCEPTION_PRINT_AND_GOTO(cleanup);
-    int64_t notAfterTime = (int64_t)(*env)->CallLongMethod(env, loc[notAfter], g_DateGetTime);
+    // byte[] bytes = new byte[] { ... }
+    // InputStream stream = new ByteArrayInputStream(bytes);
+    loc[bytes] = (*env)->NewByteArray(env, bufLen);
+    (*env)->SetByteArrayRegion(env, loc[bytes], 0, bufLen, (const jbyte *)buf);
+    loc[stream] = (*env)->NewObject(env, g_ByteArrayInputStreamClass, g_ByteArrayInputStreamCtor, loc[bytes]);
     ON_EXCEPTION_PRINT_AND_GOTO(cleanup);
 
-    // Date notBefore = cert.getNotBefore()
-    // long notBeforeTime = notBefore.getTime()
-    loc[notBefore] = (*env)->CallObjectMethod(env, cert, g_X509CertGetNotBefore);
-    ON_EXCEPTION_PRINT_AND_GOTO(cleanup);
-    int64_t notBeforeTime = (int64_t)(*env)->CallLongMethod(env, loc[notBefore], g_DateGetTime);
+    // CertificateFactory certFactory = CertificateFactory.getInstance("X.509");
+    loc[certType] = JSTRING("X.509");
+    loc[certFactory] = (*env)->CallStaticObjectMethod(env, g_CertFactoryClass, g_CertFactoryGetInstance, loc[certType]);
     ON_EXCEPTION_PRINT_AND_GOTO(cleanup);
 
-    info->Version = version;
-    info->NotAfter = notAfterTime;
-    info->NotBefore = notBeforeTime;
-    success = true;
-
-cleanup:
-    RELEASE_LOCALS(loc, env)
-    return success;
-}
-
-int32_t AndroidCryptoNative_X509GetPublicKeyAlgorithm(jobject /*X509Certificate*/ cert, char *buf, int32_t len)
-{
-    assert(cert != NULL);
-    JNIEnv *env = GetJNIEnv();
-
-    int32_t ret = BUFFER_FAIL;
-    INIT_LOCALS(loc, key, algorithm)
-
-    // PublicKey key = cert.getPublicKey();
-    // String algorithm = key.getAlgorithm();
-    // return encoded.length;
-    loc[key] = (*env)->CallObjectMethod(env, cert, g_X509CertGetPublicKey);
+    // Collection<Certificate> certs = certFactory.generateCertificates(stream);
+    loc[certs] = (*env)->CallObjectMethod(env, loc[certFactory], g_CertFactoryGenerateCertificates, loc[stream]);
     ON_EXCEPTION_PRINT_AND_GOTO(cleanup);
-    loc[algorithm] = (*env)->CallObjectMethod(env, loc[key], g_KeyGetAlgorithm);
-    ON_EXCEPTION_PRINT_AND_GOTO(cleanup);
-    ret = PopulateString(env, loc[algorithm], buf, len);
 
-cleanup:
-    RELEASE_LOCALS(loc, env)
-    return ret;
-}
-
-int32_t AndroidCryptoNative_X509GetPublicKeyBytes(jobject /*X509Certificate*/ cert, uint8_t *buf, int32_t len)
-{
-    assert(cert != NULL);
-    JNIEnv *env = GetJNIEnv();
+    jint certCount = (*env)->CallIntMethod(env, loc[certs], g_CollectionSize);
 
-    int32_t ret = BUFFER_FAIL;
-    INIT_LOCALS(loc, key, keyInfoBytes)
+    // Insufficient buffer
+    if (outLen < certCount)
+    {
+        ret = -certCount;
+        goto cleanup;
+    }
 
-    // PublicKey key = cert.getPublicKey();
-    // byte[] keyInfoBytes = key.getEncoded();
-    loc[key] = (*env)->CallObjectMethod(env, cert, g_X509CertGetPublicKey);
+    // int i = 0;
+    // Iterator<Certificate> iter = certs.iterator();
```

Comment:
remove?

Reply:
I've been trying to put the 'code' represented by the JNI calls in comments. I think it makes it a bit easier to read through.